examples/keystore/async_keystore_test.py

```diff
@@ -0,0 +1,299 @@
+  The server has a /status/{id} URL that we can query for that status.
+  The queried status is a simple string identifying the state.
+  WAITING, RUNNING, DONE (success) or ERROR (failure) where ERROR has
+  additional explanation.
+  """
+
+  @property
+  def finished(self):
+    """Implements OperationStatus API.
+
+    Indicates whether our asynchronous task has completed or not.
+    """
+    return self.__current_state in ['DONE', 'ERROR']
+
+  @property
+  def finished_ok(self):
+    """Implements OperationStatus API.
+
+    Indicates whether we finished successfully or not.
+    """
+    return self.__current_state == 'DONE'
+
+  @property
+  def error(self):
+    """Implements OperationStatus API.
+
+    Indicates whether we finished with a failure, and what the failure was.
+    """
+    return self.__error
+
+  @property
+  def id(self):
+    """The id for this status result."""
+    return self.__task_id
+
+  @property
+  def detail_path(self):
+    """The URL to query for the current task status."""
+    return self.__detail_path
+
+  @property
+  def current_state(self):
+    """A custom method to access our current task state."""
+    return self.__current_state
+
+  def __init__(self, operation, original_response=None):
+    """Constructor.
+
+    Args:
+      operation: [AgentOperation] The operation instance this is for.
+      original_response: [HttpResponseType] The response from the operation.
+    """
+    super(KeystoreStatus, self).__init__(operation, original_response)
+    self.__task_id = original_response.output
+    self.__current_state = None  # Last known state (after last refresh()).
+    self.__detail_path = '/status/{id}'.format(id=self.__task_id)
+    self.__error = None
+
+  def __str__(self):
+    return 'id={id} current_state={current_state} error={error}'.format(
+        id=self.__task_id, current_state=self.current_state,
+        error=self.__error)
+
+  def refresh(self, **kwargs):
+    """Implements HttpOperationStatus interface.
+
+    Refreshes the status state by polling the underlying server.
+    """
+    http_response = self.agent.get(self.detail_path)
+    self.set_http_response(http_response)
+
+  def set_http_response(self, http_response):
+    """Implements HttpOperationStatus interface.
+
+    Args:
+      http_response: [HttpResponseType] A response from a status poll.
+    Updates our state from a HTTP response
+    """
+    super(KeystoreStatus, self).set_http_response(http_response)
+    if http_response.http_code is None:
+      self.__current_state = 'Unknown'
+      return
+
+    parts = http_response.output.split(' ', 1)
+    self.__current_state = parts[0]
+    self.__error = parts[1] if len(parts) > 1 else None
+
+
+class AsynchronousKeystoreTestScenario(st.AgentTestScenario):
+  """Example test scenario. This is not that interesting.
+
+  It is not actually used, but we could use it to coordinate testing values
```

Comment:
Might be worth it to mention that the pattern the rest of the tests use is to define test cases in the scenario and use them in the test. Seems like we reversed that pattern a bit here.